flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -327,6 +331,9 @@ public static boolean checkpointsMatch(
      */
     void setDiscardCallback(@Nullable CompletedCheckpointStats.DiscardCallback discardCallback) {
         this.discardCallback = discardCallback;
+        if (discardCallback != null && isDiscarded) {
+            discardCallback.notifyDiscardedCheckpoint();
+        }
```

Comment:
1. It is a big question about the threading model of this class. First of all, I see that the old field `discardCallback` was volatile. Secondly, I see that `discard` method can be called from `CheckpointsCleaner` or `CheckpointStore` which can be invoked from ExecutionThreadPool.
2. Triggering the callback is required because of what I described above if that instance can be used in different threads we should be sure that the callback will be called at least once(right now it is possible that it can be called twice). 

In general, I agree that the current implementation is more complicated. But I didn't find an easier way. We can try to create `CompletedCheckpointStats` in the same place where we did it before the changes but then we need somewhere to hold it before reporting to `tracker`. I mean right now we do the reporting here `PendingCheckpointStats#reportCompletedCheckpoint` but if we decided to separate this method into two(creating, reporting) then we need to store created object somewhere(CompletedCheckpoint maybe).

Reply:
> right now it is possible that it can be called twice

Aren't all of those places mutually exclusive? That logic in the `CheckpointCoordinator` guarantees that only one of them will be executed? I'm pretty sure that at the very least discard and non-discard code paths must be mutually exclusive.

Frankly it sounds like `notifyDiscardedCheckpoint()` should be more something like `markCheckpointDiscarded()` called from the `CheckpointsCoordinator`, around the same place that checkpoint is decided to be discarded, and not where checkpoint is being actually discarded?

Reply:
I don't sure that I fully follow your idea but for example line `CheckpointCoordinator:1385`, if we fail with the subsume we discard the checkpoint asynchronously(at least according to comment `executor` is async):
```
// we failed to store the completed checkpoint. Let's clean up
checkpointsCleaner.cleanCheckpointOnFailedStoring(completedCheckpoint, executor);
```
So I don't see any protection here.